# Análisis de datos con Spark

In [0]:
# aunque en la guía decia cargarlo como volumen, lo cargamos como tablas. Solo de esta manera logramos importar los csv

vehiculos_involucrados = spark.table("default.vehiculos_involucrados_corregido")
fallecidos_lesionados = spark.table("default.fallecidos_lesionados_corregido")
hechos_transito = spark.table("default.hechos_transito_corregido")

#1 Mostrar cuantos registros hay en cada uno
print("vehiculos_involucrados: ", vehiculos_involucrados.count())
print("hechos_transito: ", hechos_transito.count())
print("fallecidos_lesionados: ", fallecidos_lesionados.count()) 

# y usar la funcion .show para ver algunos resultados de cada uno
vehiculos_involucrados.show(5)
print()
hechos_transito.show(5)
print()
fallecidos_lesionados.show(5)


vehiculos_involucrados:  111988
hechos_transito:  76759
fallecidos_lesionados:  109417
+---------+-------+-------+--------+---------+---------+---------+--------+--------+---------+----------+---------+---------+----------+--------+--------+--------------------+--------------------+----+
|num_hecho|dia_ocu|mes_ocu|hora_ocu|depto_ocu|mupio_ocu|areag_ocu|zona_ocu|tipo_veh|color_veh|modelo_veh|causa_acc|marca_veh|estado_pil|sexo_pil|edad_pil|      archivo_origen|        tipo_dataset| año|
+---------+-------+-------+--------+---------+---------+---------+--------+--------+---------+----------+---------+---------+----------+--------+--------+--------------------+--------------------+----+
|      1.0|    1.0|    1.0|     3.0|      1.0|    106.0|      1.0|     6.0|     3.0|     99.0|       6.0|      2.0|     99.0|       2.0|     1.0|    32.0|vehiculos_involuc...|vehiculos_involuc...|2013|
|      2.0|    1.0|    1.0|     5.0|      1.0|    101.0|      1.0|    11.0|     1.0|      2.0|       5.0|

In [0]:
# Hacer un summary para las columnas importantes de cada uno
# Versión compacta del summary
print("\nVARIABLES PRINCIPALES DE CADA ARCHIVO:")

print("\nVEHÍCULOS INVOLUCRADOS - Summary:")
vehiculos_summary = vehiculos_involucrados.select(
    "edad_pil", "mes_ocu", "hora_ocu"
).summary("count", "min", "max")
vehiculos_summary.show()

print("\nHECHOS DE TRÁNSITO - Summary:")
hechos_summary = hechos_transito.select(
    "mes_ocu", "hora_ocu", "dia_ocu", "dia_sem_ocu" 
).summary("count", "min", "max")
hechos_summary.show()

print("\nFALLECIDOS Y LESIONADOS - Summary:")
fallecidos_summary = fallecidos_lesionados.select(
    "edad_pil", "mes_ocu", "hora_ocu", "tipo_eve"
).summary("count", "min", "max")
fallecidos_summary.show()


VARIABLES PRINCIPALES DE CADA ARCHIVO:

VEHÍCULOS INVOLUCRADOS - Summary:
+-------+--------+-------+--------+
|summary|edad_pil|mes_ocu|hora_ocu|
+-------+--------+-------+--------+
|  count|  111988| 111988|  111988|
|    min|     3.0|    1.0|     0.0|
|    max|   999.0|   12.0|    99.0|
+-------+--------+-------+--------+


HECHOS DE TRÁNSITO - Summary:
+-------+-------+--------+-------+-----------+
|summary|mes_ocu|hora_ocu|dia_ocu|dia_sem_ocu|
+-------+-------+--------+-------+-----------+
|  count|  76759|   76759|  76759|      76759|
|    min|    1.0|     0.0|    1.0|        1.0|
|    max|   12.0|    99.0|   31.0|        7.0|
+-------+-------+--------+-------+-----------+


FALLECIDOS Y LESIONADOS - Summary:
+-------+--------+-------+--------+--------+
|summary|edad_pil|mes_ocu|hora_ocu|tipo_eve|
+-------+--------+-------+--------+--------+
|  count|  100427| 109417|  109417|  100357|
|    min|     0.0|    1.0|     0.0|     1.0|
|    max|   999.0|   12.0|    99.0|    99.0|
+----

Podemos ver que si se han cargado los datos de manera exitosa usando spark desde databricks, tuvimos que cargar los csv como tablas, pero aparte de eso no hemos hecho mayores cambios. Primero podemos ver algunas estadísticas usando summary donde podemos ver cosas como hora de ocurrencia, mes de ocurrencia entre otros. Algo interesante a mencionar con estos resultados, es que se puede ver que hay valores muy altos para la edad de los pilotos. Esto no necesariamente es un error del script, de hecho de acuerdo con el libro de variables que se puede encontrar en la página del INE, esto hace sentido. Las personas que ingresan los datos tienden a colocar valores muy altos en casillas numéricas cuando estas son desconocidas. Se puede ver en un caso que el maximo es 999, lo cuál es un claro indicio de que así se marcan valores numéricos desconocidos. Esto será importante tenerlo en cuenta en siguientes pasos. Similar a como funciona summary, es posible usar describe para estadísticas numéricas más completas sin necesidad de especificarlas.

In [0]:
print("\nVEHÍCULOS - Describe:")
print("Numéricas:")
vehiculos_involucrados.select("edad_pil", "mes_ocu", "hora_ocu").describe().show()
print("\nCategóricas:")
hechos_transito.select("mes_ocu", "hora_ocu", "dia_ocu").describe().show()
print("Numéricas:")
fallecidos_lesionados.select("edad_pil", "mes_ocu", "hora_ocu").describe().show()



VEHÍCULOS - Describe:
Numéricas:
+-------+------------------+-----------------+------------------+
|summary|          edad_pil|          mes_ocu|          hora_ocu|
+-------+------------------+-----------------+------------------+
|  count|            111988|           111988|            111988|
|   mean| 256.2040040004286|6.597055041611601|13.702557416866092|
| stddev|407.41133258885975| 3.51955144121383| 7.020649994962808|
|    min|               3.0|              1.0|               0.0|
|    max|             999.0|             12.0|              99.0|
+-------+------------------+-----------------+------------------+


Categóricas:
+-------+------------------+------------------+------------------+
|summary|           mes_ocu|          hora_ocu|           dia_ocu|
+-------+------------------+------------------+------------------+
|  count|             76759|             76759|             76759|
|   mean|6.5776781875740955|13.719928607720266|15.616123190765904|
| stddev| 3.5268926483

In [0]:
fallecidos_lesionados.printSchema()
hechos_transito.printSchema()
vehiculos_involucrados.printSchema()

root
 |-- num_hecho: double (nullable = true)
 |-- dia_ocu: double (nullable = true)
 |-- mes_ocu: double (nullable = true)
 |-- hora_ocu: double (nullable = true)
 |-- depto_ocu: double (nullable = true)
 |-- mupio_ocu: double (nullable = true)
 |-- areag_ocu: double (nullable = true)
 |-- zona_ocu: double (nullable = true)
 |-- tipo_veh: double (nullable = true)
 |-- causa_acc: double (nullable = true)
 |-- sexo_pil: double (nullable = true)
 |-- edad_pil: double (nullable = true)
 |-- fallecidos_lesionados: double (nullable = true)
 |-- tipo_eve: double (nullable = true)
 |-- archivo_origen: string (nullable = true)
 |-- tipo_dataset: string (nullable = true)
 |-- año: long (nullable = true)

root
 |-- num_hecho: double (nullable = true)
 |-- dia_ocu: double (nullable = true)
 |-- mes_ocu: double (nullable = true)
 |-- dia_sem_ocu: double (nullable = true)
 |-- hora_ocu: double (nullable = true)
 |-- depto_ocu: double (nullable = true)
 |-- mupio_ocu: double (nullable = true)
 |-- a

Con estos resultados nuevamente se ve que hay números grandes y la media y desviación estandar muestran que puede que hayan varios valores atípicos dentro del set de datos.Además de tener en cuenta que varios valores son numéricos, debido a que riginalmente eran solo ID's en varios casos.

# Validación de años

Tomando en cuenta que todos estos datos provienen de distintos archivos es importante verificar que si tengamos correctamente el rango de los años. En el pre procesamiento de datos si alguno no tenía año se lo hemos colocado dependiendo del archivo de origen, por lo que debería de estar el rango completo.

In [0]:
fallecidos_lesionados.select("año").distinct().show()
hechos_transito.select("año").distinct().show()
vehiculos_involucrados.select("año").distinct().show()

+----+
| año|
+----+
|2013|
|2014|
|2015|
|2016|
|2017|
|2018|
|2019|
|2020|
|2021|
|2022|
|2023|
+----+

+----+
| año|
+----+
|2013|
|2014|
|2015|
|2016|
|2017|
|2018|
|2019|
|2020|
|2021|
|2022|
|2023|
+----+

+----+
| año|
+----+
|2013|
|2014|
|2015|
|2016|
|2017|
|2018|
|2019|
|2020|
|2021|
|2022|
|2023|
+----+



Como se puede ver todos cumplen con el ranfo de 2013 hasta 2023.

In [0]:
# mostrar los distintos tipos de accidentes
fallecidos_lesionados.select("tipo_eve").distinct().show()
hechos_transito.select("causa_acc").distinct().show()
vehiculos_involucrados.select("causa_acc").distinct().show()

+--------+
|tipo_eve|
+--------+
|     1.0|
|     3.0|
|     4.0|
|     2.0|
|     5.0|
|    99.0|
|     7.0|
|     8.0|
|     6.0|
|    NULL|
+--------+

+---------+
|causa_acc|
+---------+
|      2.0|
|      1.0|
|      3.0|
|      4.0|
|     99.0|
|      5.0|
|     NULL|
+---------+

+---------+
|causa_acc|
+---------+
|      2.0|
|      1.0|
|      3.0|
|      4.0|
|     99.0|
|     NULL|
+---------+



Estos valores son números porque seguramente dentro de los respectivos archivos eran ID's. 

In [0]:
from pyspark.sql.functions import when, col

# Enfoque: Crear nuevas columnas y luego eliminar las originales

# Para fallecidos_lesionados - tipo_eve
fallecidos_lesionados = fallecidos_lesionados.withColumn(
    "tipo_eve_desc",
    when(col("tipo_eve") == 1, "Colisión")
    .when(col("tipo_eve") == 2, "Choque")
    .when(col("tipo_eve") == 3, "Vuelco")
    .when(col("tipo_eve") == 4, "Caída")
    .when(col("tipo_eve") == 5, "Atropello")
    .when(col("tipo_eve") == 6, "Pérdida de control")
    .when(col("tipo_eve") == 7, "Colisión contra animal")
    .when(col("tipo_eve") == 8, "Exceso de pasaje")
    .when(col("tipo_eve") == 9, "Asfalto mojado")
    .when(col("tipo_eve") == 10, "Exceso de velocidad")
    .when(col("tipo_eve") == 11, "Desperfectos mecánicos")
    .when(col("tipo_eve") == 12, "Incendio")
    .when(col("tipo_eve") == 99, "Ignorado")
    .otherwise("Otro")
)

# Para hechos_transito - causa_acc
hechos_transito = hechos_transito.withColumn(
    "causa_acc_desc",
    when(col("causa_acc") == 1, "Colisión")
    .when(col("causa_acc") == 2, "Choque")
    .when(col("causa_acc") == 3, "Vuelco")
    .when(col("causa_acc") == 4, "Caída")
    .when(col("causa_acc") == 5, "Atropello")
    .when(col("causa_acc") == 6, "Derrape")
    .when(col("causa_acc") == 7, "Embarranco")
    .when(col("causa_acc") == 8, "Encuneto")
    .when(col("causa_acc") == 99, "Ignorado")
    .otherwise("Otro")
)

# Para vehiculos_involucrados - causa_acc
vehiculos_involucrados = vehiculos_involucrados.withColumn(
    "causa_acc_desc",
    when(col("causa_acc") == 1, "Colisión")
    .when(col("causa_acc") == 2, "Choque")
    .when(col("causa_acc") == 3, "Vuelco")
    .when(col("causa_acc") == 4, "Caída")
    .when(col("causa_acc") == 5, "Atropello")
    .when(col("causa_acc") == 6, "Derrape")
    .when(col("causa_acc") == 7, "Embarranco")
    .when(col("causa_acc") == 8, "Encuneto")
    .when(col("causa_acc") == 99, "Ignorado")
    .otherwise("Otro")
)


In [0]:
print("FALLECIDOS_LESIONADOS - Tipos de evento únicos:")
fallecidos_lesionados.select("tipo_eve_desc").distinct().show()

print("HECHOS_TRANSITO - Causas de accidente únicas:")
hechos_transito.select("causa_acc_desc").distinct().show()

print("VEHICULOS_INVOLUCRADOS - Causas de accidente únicas:")
vehiculos_involucrados.select("causa_acc_desc").distinct().show()


FALLECIDOS_LESIONADOS - Tipos de evento únicos:
+--------------------+
|       tipo_eve_desc|
+--------------------+
|                Otro|
|            Colisión|
|              Vuelco|
|               Caída|
|              Choque|
|           Atropello|
|            Ignorado|
|Colisión contra a...|
|    Exceso de pasaje|
|  Pérdida de control|
+--------------------+

HECHOS_TRANSITO - Causas de accidente únicas:
+--------------+
|causa_acc_desc|
+--------------+
|        Choque|
|      Colisión|
|        Vuelco|
|         Caída|
|      Ignorado|
|     Atropello|
|          Otro|
+--------------+

VEHICULOS_INVOLUCRADOS - Causas de accidente únicas:
+--------------+
|causa_acc_desc|
+--------------+
|        Choque|
|      Colisión|
|        Vuelco|
|         Caída|
|      Ignorado|
|          Otro|
+--------------+



Con esto se puede ver los distintos tipos de eventos y accidentes por cada uno de los sets de datos, colocando una columna nueva con el valor del tipo de accidente. Los valores fueron extaidos directamente del diccionario de varaibles del ultimo año usado, 2023.

# Departamentos por set de datos

In [0]:
# Departamentos unicos por set de datos, de igual manera es una columna numerica entonces hay que reemplazarlo por sus valores reales.

from pyspark.sql.functions import when, col

# Aplicar la transformación para departamentos en los tres datasets

# Para fallecidos_lesionados - depto_ocu
fallecidos_lesionados = fallecidos_lesionados.withColumn(
    "depto_ocu_desc",
    when(col("depto_ocu") == 1, "Guatemala")
    .when(col("depto_ocu") == 2, "El progreso")
    .when(col("depto_ocu") == 3, "Sacatepéquez")
    .when(col("depto_ocu") == 4, "Chimaltenango")
    .when(col("depto_ocu") == 5, "Escuintla")
    .when(col("depto_ocu") == 6, "Santa rosa")
    .when(col("depto_ocu") == 7, "Sololá")
    .when(col("depto_ocu") == 8, "Totonicapán")
    .when(col("depto_ocu") == 9, "Quetzaltenango")
    .when(col("depto_ocu") == 10, "Suchitepéquez")
    .when(col("depto_ocu") == 11, "Retalhuleu")
    .when(col("depto_ocu") == 12, "San Marcos")
    .when(col("depto_ocu") == 13, "Huehuetenango")
    .when(col("depto_ocu") == 14, "Quiche")
    .when(col("depto_ocu") == 15, "Baja Verapaz")
    .when(col("depto_ocu") == 16, "Alta Verapaz")
    .when(col("depto_ocu") == 17, "Peten")
    .when(col("depto_ocu") == 18, "Izabal")
    .when(col("depto_ocu") == 19, "Zacapa")
    .when(col("depto_ocu") == 20, "Chiquimula")
    .when(col("depto_ocu") == 21, "Jalapa")
    .when(col("depto_ocu") == 22, "Jutiapa")
    .otherwise("Desconocido")
)

# Para hechos_transito - depto_ocu
hechos_transito = hechos_transito.withColumn(
    "depto_ocu_desc",
    when(col("depto_ocu") == 1, "Guatemala")
    .when(col("depto_ocu") == 2, "El progreso")
    .when(col("depto_ocu") == 3, "Sacatepéquez")
    .when(col("depto_ocu") == 4, "Chimaltenango")
    .when(col("depto_ocu") == 5, "Escuintla")
    .when(col("depto_ocu") == 6, "Santa rosa")
    .when(col("depto_ocu") == 7, "Sololá")
    .when(col("depto_ocu") == 8, "Totonicapán")
    .when(col("depto_ocu") == 9, "Quetzaltenango")
    .when(col("depto_ocu") == 10, "Suchitepéquez")
    .when(col("depto_ocu") == 11, "Retalhuleu")
    .when(col("depto_ocu") == 12, "San Marcos")
    .when(col("depto_ocu") == 13, "Huehuetenango")
    .when(col("depto_ocu") == 14, "Quiche")
    .when(col("depto_ocu") == 15, "Baja Verapaz")
    .when(col("depto_ocu") == 16, "Alta Verapaz")
    .when(col("depto_ocu") == 17, "Peten")
    .when(col("depto_ocu") == 18, "Izabal")
    .when(col("depto_ocu") == 19, "Zacapa")
    .when(col("depto_ocu") == 20, "Chiquimula")
    .when(col("depto_ocu") == 21, "Jalapa")
    .when(col("depto_ocu") == 22, "Jutiapa")
    .otherwise("Desconocido")
)

# Para vehiculos_involucrados - depto_ocu
vehiculos_involucrados = vehiculos_involucrados.withColumn(
    "depto_ocu_desc",
    when(col("depto_ocu") == 1, "Guatemala")
    .when(col("depto_ocu") == 2, "El progreso")
    .when(col("depto_ocu") == 3, "Sacatepéquez")
    .when(col("depto_ocu") == 4, "Chimaltenango")
    .when(col("depto_ocu") == 5, "Escuintla")
    .when(col("depto_ocu") == 6, "Santa rosa")
    .when(col("depto_ocu") == 7, "Sololá")
    .when(col("depto_ocu") == 8, "Totonicapán")
    .when(col("depto_ocu") == 9, "Quetzaltenango")
    .when(col("depto_ocu") == 10, "Suchitepéquez")
    .when(col("depto_ocu") == 11, "Retalhuleu")
    .when(col("depto_ocu") == 12, "San Marcos")
    .when(col("depto_ocu") == 13, "Huehuetenango")
    .when(col("depto_ocu") == 14, "Quiche")
    .when(col("depto_ocu") == 15, "Baja Verapaz")
    .when(col("depto_ocu") == 16, "Alta Verapaz")
    .when(col("depto_ocu") == 17, "Peten")
    .when(col("depto_ocu") == 18, "Izabal")
    .when(col("depto_ocu") == 19, "Zacapa")
    .when(col("depto_ocu") == 20, "Chiquimula")
    .when(col("depto_ocu") == 21, "Jalapa")
    .when(col("depto_ocu") == 22, "Jutiapa")
    .otherwise("Desconocido")
)


In [0]:
# ver lo nuevo
print("FALLECIDOS_LESIONADOS - Departamento:")
fallecidos_lesionados.select("depto_ocu_desc").distinct().show()
print("HECHOS_TRANSITO - Departamento:")
hechos_transito.select("depto_ocu_desc").distinct().show()
print("VEHICULOS_INVOLUCRADOS - Departamento:")
vehiculos_involucrados.select("depto_ocu_desc").distinct().show()

# mostrar el largo de las columnas depto_ocu_desc
print("FALLECIDOS_LESIONADOS - Departamento:")
print(fallecidos_lesionados.select("depto_ocu_desc").distinct().count())

print("HECHOS_TRANSITO - Departamento:")
print(hechos_transito.select("depto_ocu_desc").distinct().count())

print("VEHICULOS_INVOLUCRADOS - Departamento:")
print(vehiculos_involucrados.select("depto_ocu_desc").distinct().count())

FALLECIDOS_LESIONADOS - Departamento:
+--------------+
|depto_ocu_desc|
+--------------+
|     Guatemala|
|        Zacapa|
|     Escuintla|
|    San Marcos|
| Huehuetenango|
|        Izabal|
|  Sacatepéquez|
|  Alta Verapaz|
|    Santa rosa|
|Quetzaltenango|
|   El progreso|
|        Sololá|
|    Chiquimula|
| Suchitepéquez|
|  Baja Verapaz|
|   Totonicapán|
|        Quiche|
| Chimaltenango|
|       Jutiapa|
|         Peten|
+--------------+
only showing top 20 rows
HECHOS_TRANSITO - Departamento:
+--------------+
|depto_ocu_desc|
+--------------+
|     Guatemala|
|        Zacapa|
|     Escuintla|
|Quetzaltenango|
|    San Marcos|
| Huehuetenango|
|  Alta Verapaz|
|        Izabal|
|  Sacatepéquez|
|    Santa rosa|
|   El progreso|
|        Sololá|
|    Chiquimula|
| Suchitepéquez|
|  Baja Verapaz|
|    Retalhuleu|
|   Totonicapán|
|        Quiche|
| Chimaltenango|
|       Jutiapa|
+--------------+
only showing top 20 rows
VEHICULOS_INVOLUCRADOS - Departamento:
+--------------+
|depto_o

Podemos ver que todos los departamentos tienen al menos un evento, accidente o similar rpeortado en cada uno de las tablas. Y hemos hechp una columna depto_occu_desc para poder tener el valor de string y non solo el ID.

In [0]:
from pyspark.sql import functions as F

print("Número de accidentes por año y departamento - FALLECIDOS_LESIONADOS")
df_fallecidos_agrupado = fallecidos_lesionados.groupBy("año", "depto_ocu_desc").count().orderBy("año", "depto_ocu_desc")
display(df_fallecidos_agrupado)

print("Número de accidentes por año y departamento - HECHOS_TRANSITO")
df_hechos_agrupado = hechos_transito.groupBy("año", "depto_ocu_desc").count().orderBy("año", "depto_ocu_desc")
display(df_hechos_agrupado)

print("Número de accidentes por año y departamento - VEHICULOS_INVOLUCRADOS")
df_vehiculos_agrupado = vehiculos_involucrados.groupBy("año", "depto_ocu_desc").count().orderBy("año", "depto_ocu_desc")
display(df_vehiculos_agrupado)

Número de accidentes por año y departamento - FALLECIDOS_LESIONADOS


año depto_ocu_desc count 2013 Alta Verapaz 317 2013 Baja Verapaz 191 2013 Chimaltenango 401 2013 Chiquimula 271 2013 El progreso 200 2013 Escuintla 675 2013 Guatemala 2648 2013 Huehuetenango 256 2013 Izabal 310 2013 Jalapa 138 2013 Jutiapa 284 2013 Peten 308 2013 Quetzaltenango 505 2013 Quiche 321 2013 Retalhuleu 208 2013 Sacatepéquez 272 2013 San Marcos 409 2013 Santa rosa 320 2013 Sololá 193 2013 Suchitepéquez 410 2013 Totonicapán 109 2013 Zacapa 314 2014 Alta Verapaz 428 2014 Baja Verapaz 158 2014 Chimaltenango 392 2014 Chiquimula 205 2014 El progreso 220 2014 Escuintla 741 2014 Guatemala 2503 2014 Huehuetenango 280 2014 Izabal 376 2014 Jalapa 105 2014 Jutiapa 288 2014 Peten 348 2014 Quetzaltenango 350 2014 Quiche 363 2014 Retalhuleu 227 2014 Sacatepéquez 282 2014 San Marcos 266 2014 Santa rosa 413 2014 Sololá 268 2014 Suchitepéquez 332 2014 Totonicapán 147 2014 Zacapa 298 2015 Alta Verapaz 607 2015 Baja Verapaz 213 2015 Chimaltenango 480 2015 Chiquimula 284 2015 El progreso 255 2015 Escuintla 803 2015 Guatemala 2947 2015 Huehuetenango 261 2015 Izabal 425 2015 Jalapa 221 2015 Jutiapa 333 2015 Peten 456 2015 Quetzaltenango 401 2015 Quiche 361 2015 Retalhuleu 312 2015 Sacatepéquez 296 2015 San Marcos 298 2015 Santa rosa 381 2015 Sololá 207 2015 Suchitepéquez 339 2015 Totonicapán 137 2015 Zacapa 380 2016 Alta Verapaz 646 2016 Baja Verapaz 271 2016 Chimaltenango 590 2016 Chiquimula 283 2016 El progreso 318 2016 Escuintla 958 2016 Guatemala 3255 2016 Huehuetenango 363 2016 Izabal 489 2016 Jalapa 157 2016 Jutiapa 278 2016 Peten 411 2016 Quetzaltenango 438 2016 Quiche 330 2016 Retalhuleu 327 2016 Sacatepéquez 468 2016 San Marcos 275 2016 Santa rosa 441 2016 Sololá 398 2016 Suchitepéquez 441 2016 Totonicapán 110 2016 Zacapa 421 2017 Alta Verapaz 411 2017 Baja Verapaz 161 2017 Chimaltenango 368 2017 Chiquimula 162 2017 El progreso 251 2017 Escuintla 758 2017 Guatemala 2711 2017 Huehuetenango 297 2017 Izabal 282 2017 Jalapa 137 2017 Jutiapa 303 2017 Peten 331 2017 Quetzaltenango 220 2017 Quiche 283 2017 Retalhuleu 197 2017 Sacatepéquez 331 2017 San Marcos 184 2017 Santa rosa 420 2017 Sololá 186 2017 Suchitepéquez 329 2017 Totonicapán 66 2017 Zacapa 237 2018 Alta Verapaz 567 2018 Baja Verapaz 217 2018 Chimaltenango 363 2018 Chiquimula 257 2018 El progreso 275 2018 Escuintla 679 2018 Guatemala 3459 2018 Huehuetenango 210 2018 Izabal 353 2018 Jalapa 120 2018 Jutiapa 246 2018 Peten 251 2018 Quetzaltenango 230 2018 Quiche 222 2018 Retalhuleu 229 2018 Sacatepéquez 296 2018 San Marcos 269 2018 Santa rosa 290 2018 Sololá 166 2018 Suchitepéquez 329 2018 Totonicapán 77 2018 Zacapa 302 2019 Alta Verapaz 583 2019 Baja Verapaz 244 2019 Chimaltenango 400 2019 Chiquimula 262 2019 El progreso 298 2019 Escuintla 838 2019 Guatemala 3805 2019 Huehuetenango 165 2019 Izabal 358 2019 Jalapa 125 2019 Jutiapa 365 2019 Peten 367 2019 Quetzaltenango 340 2019 Quiche 180 2019 Retalhuleu 308 2019 Sacatepéquez 394 2019 San Marcos 222 2019 Santa rosa 349 2019 Sololá 286 2019 Suchitepéquez 385 2019 Totonicapán 120 2019 Zacapa 270 2020 Alta Verapaz 400 2020 Baja Verapaz 165 2020 Chimaltenango 316 2020 Chiquimula 205 2020 El progreso 198 2020 Escuintla 659 2020 Guatemala 3025 2020 Huehuetenango 173 2020 Izabal 294 2020 Jalapa 162 2020 Jutiapa 224 2020 Peten 293 2020 Quetzaltenango 240 2020 Quiche 172 2020 Retalhuleu 255 2020 Sacatepéquez 262 2020 San Marcos 188 2020 Santa rosa 283 2020 Sololá 101 2020 Suchitepéquez 256 2020 Totonicapán 74 2020 Zacapa 197 2021 Alta Verapaz 607 2021 Baja Verapaz 183 2021 Chimaltenango 372 2021 Chiquimula 261 2021 El progreso 253 2021 Escuintla 896 2021 Guatemala 3982 2021 Huehuetenango 196 2021 Izabal 332 2021 Jalapa 209 2021 Jutiapa 326 2021 Peten 368 2021 Quetzaltenango 296 2021 Quiche 302 2021 Retalhuleu 284 2021 Sacatepéquez 336 2021 San Marcos 217 2021 Santa rosa 333 2021 Sololá 177 2021 Suchitepéquez 293 2021 Totonicapán 79 2021 Zacapa 242 2022 Alta Verapaz 518 2022 Baja Verapaz 250 2022 Chimaltenango 403 2022 Chiquimula 236 2022

Databricks visualization. Run in Databricks to view.

Número de accidentes por año y departamento - HECHOS_TRANSITO


año depto_ocu_desc count 2013 Alta Verapaz 219 2013 Baja Verapaz 100 2013 Chimaltenango 212 2013 Chiquimula 204 2013 El progreso 132 2013 Escuintla 424 2013 Guatemala 2267 2013 Huehuetenango 201 2013 Izabal 234 2013 Jalapa 89 2013 Jutiapa 174 2013 Peten 214 2013 Quetzaltenango 288 2013 Quiche 151 2013 Retalhuleu 133 2013 Sacatepéquez 142 2013 San Marcos 221 2013 Santa rosa 214 2013 Sololá 126 2013 Suchitepéquez 269 2013 Totonicapán 81 2013 Zacapa 229 2014 Alta Verapaz 223 2014 Baja Verapaz 95 2014 Chimaltenango 228 2014 Chiquimula 126 2014 El progreso 127 2014 Escuintla 444 2014 Guatemala 1912 2014 Huehuetenango 136 2014 Izabal 255 2014 Jalapa 75 2014 Jutiapa 172 2014 Peten 219 2014 Quetzaltenango 255 2014 Quiche 144 2014 Retalhuleu 132 2014 Sacatepéquez 150 2014 San Marcos 154 2014 Santa rosa 242 2014 Sololá 109 2014 Suchitepéquez 223 2014 Totonicapán 82 2014 Zacapa 148 2015 Alta Verapaz 315 2015 Baja Verapaz 112 2015 Chimaltenango 263 2015 Chiquimula 172 2015 El progreso 160 2015 Escuintla 569 2015 Guatemala 2273 2015 Huehuetenango 173 2015 Izabal 285 2015 Jalapa 96 2015 Jutiapa 182 2015 Peten 288 2015 Quetzaltenango 316 2015 Quiche 157 2015 Retalhuleu 181 2015 Sacatepéquez 209 2015 San Marcos 177 2015 Santa rosa 254 2015 Sololá 152 2015 Suchitepéquez 243 2015 Totonicapán 69 2015 Zacapa 208 2016 Alta Verapaz 353 2016 Baja Verapaz 149 2016 Chimaltenango 420 2016 Chiquimula 181 2016 El progreso 201 2016 Escuintla 627 2016 Guatemala 2619 2016 Huehuetenango 225 2016 Izabal 286 2016 Jalapa 98 2016 Jutiapa 180 2016 Peten 267 2016 Quetzaltenango 323 2016 Quiche 195 2016 Retalhuleu 262 2016 Sacatepéquez 287 2016 San Marcos 198 2016 Santa rosa 261 2016 Sololá 182 2016 Suchitepéquez 259 2016 Totonicapán 76 2016 Zacapa 315 2017 Alta Verapaz 247 2017 Baja Verapaz 111 2017 Chimaltenango 244 2017 Chiquimula 114 2017 El progreso 139 2017 Escuintla 461 2017 Guatemala 2128 2017 Huehuetenango 140 2017 Izabal 200 2017 Jalapa 92 2017 Jutiapa 177 2017 Peten 194 2017 Quetzaltenango 172 2017 Quiche 145 2017 Retalhuleu 148 2017 Sacatepéquez 213 2017 San Marcos 130 2017 Santa rosa 234 2017 Sololá 124 2017 Suchitepéquez 238 2017 Totonicapán 44 2017 Zacapa 184 2018 Alta Verapaz 265 2018 Baja Verapaz 132 2018 Chimaltenango 238 2018 Chiquimula 193 2018 El progreso 168 2018 Escuintla 407 2018 Guatemala 2722 2018 Huehuetenango 108 2018 Izabal 210 2018 Jalapa 78 2018 Jutiapa 160 2018 Peten 187 2018 Quetzaltenango 155 2018 Quiche 139 2018 Retalhuleu 153 2018 Sacatepéquez 201 2018 San Marcos 147 2018 Santa rosa 192 2018 Sololá 104 2018 Suchitepéquez 205 2018 Totonicapán 47 2018 Zacapa 184 2019 Alta Verapaz 317 2019 Baja Verapaz 129 2019 Chimaltenango 274 2019 Chiquimula 153 2019 El progreso 151 2019 Escuintla 452 2019 Guatemala 3017 2019 Huehuetenango 99 2019 Izabal 234 2019 Jalapa 94 2019 Jutiapa 219 2019 Peten 217 2019 Quetzaltenango 191 2019 Quiche 114 2019 Retalhuleu 207 2019 Sacatepéquez 233 2019 San Marcos 149 2019 Santa rosa 220 2019 Sololá 140 2019 Suchitepéquez 218 2019 Totonicapán 63 2019 Zacapa 156 2020 Alta Verapaz 296 2020 Baja Verapaz 96 2020 Chimaltenango 229 2020 Chiquimula 130 2020 El progreso 124 2020 Escuintla 440 2020 Guatemala 2676 2020 Huehuetenango 113 2020 Izabal 230 2020 Jalapa 121 2020 Jutiapa 165 2020 Peten 206 2020 Quetzaltenango 197 2020 Quiche 106 2020 Retalhuleu 203 2020 Sacatepéquez 202 2020 San Marcos 147 2020 Santa rosa 197 2020 Sololá 81 2020 Suchitepéquez 201 2020 Totonicapán 49 2020 Zacapa 141 2021 Alta Verapaz 380 2021 Baja Verapaz 137 2021 Chimaltenango 297 2021 Chiquimula 166 2021 El progreso 170 2021 Escuintla 615 2021 Guatemala 3450 2021 Huehuetenango 141 2021 Izabal 237 2021 Jalapa 146 2021 Jutiapa 228 2021 Peten 266 2021 Quetzaltenango 253 2021 Quiche 186 2021 Retalhuleu 222 2021 Sacatepéquez 270 2021 San Marcos 182 2021 Santa rosa 238 2021 Sololá 109 2021 Suchitepéquez 229 2021 Totonicapán 77 2021 Zacapa 154 2022 Alta Verapaz 355 2022 Baja Verapaz 145 2022 Chimaltenango 297 2022 Chiquimula 135 2022 El progreso 147

Databricks visualization. Run in Databricks to view.

Número de accidentes por año y departamento - VEHICULOS_INVOLUCRADOS


año depto_ocu_desc count 2013 Alta Verapaz 162 2013 Baja Verapaz 109 2013 Chimaltenango 143 2013 Chiquimula 161 2013 El progreso 152 2013 Escuintla 494 2013 Guatemala 2423 2013 Huehuetenango 177 2013 Izabal 230 2013 Jalapa 92 2013 Jutiapa 168 2013 Peten 195 2013 Quetzaltenango 333 2013 Quiche 143 2013 Retalhuleu 146 2013 Sacatepéquez 136 2013 San Marcos 194 2013 Santa rosa 205 2013 Sololá 93 2013 Suchitepéquez 278 2013 Totonicapán 62 2013 Zacapa 227 2014 Alta Verapaz 289 2014 Baja Verapaz 134 2014 Chimaltenango 296 2014 Chiquimula 170 2014 El progreso 178 2014 Escuintla 610 2014 Guatemala 2775 2014 Huehuetenango 183 2014 Izabal 370 2014 Jalapa 111 2014 Jutiapa 230 2014 Peten 308 2014 Quetzaltenango 385 2014 Quiche 178 2014 Retalhuleu 181 2014 Sacatepéquez 211 2014 San Marcos 206 2014 Santa rosa 326 2014 Sololá 139 2014 Suchitepéquez 294 2014 Totonicapán 121 2014 Zacapa 209 2015 Alta Verapaz 422 2015 Baja Verapaz 158 2015 Chimaltenango 380 2015 Chiquimula 227 2015 El progreso 240 2015 Escuintla 817 2015 Guatemala 3361 2015 Huehuetenango 212 2015 Izabal 408 2015 Jalapa 137 2015 Jutiapa 255 2015 Peten 404 2015 Quetzaltenango 476 2015 Quiche 207 2015 Retalhuleu 274 2015 Sacatepéquez 298 2015 San Marcos 250 2015 Santa rosa 353 2015 Sololá 188 2015 Suchitepéquez 353 2015 Totonicapán 98 2015 Zacapa 305 2016 Alta Verapaz 470 2016 Baja Verapaz 216 2016 Chimaltenango 579 2016 Chiquimula 249 2016 El progreso 296 2016 Escuintla 961 2016 Guatemala 3904 2016 Huehuetenango 295 2016 Izabal 420 2016 Jalapa 149 2016 Jutiapa 260 2016 Peten 390 2016 Quetzaltenango 485 2016 Quiche 272 2016 Retalhuleu 396 2016 Sacatepéquez 403 2016 San Marcos 294 2016 Santa rosa 384 2016 Sololá 228 2016 Suchitepéquez 394 2016 Totonicapán 114 2016 Zacapa 459 2017 Alta Verapaz 336 2017 Baja Verapaz 156 2017 Chimaltenango 332 2017 Chiquimula 174 2017 El progreso 213 2017 Escuintla 678 2017 Guatemala 3201 2017 Huehuetenango 195 2017 Izabal 297 2017 Jalapa 132 2017 Jutiapa 261 2017 Peten 297 2017 Quetzaltenango 252 2017 Quiche 193 2017 Retalhuleu 225 2017 Sacatepéquez 308 2017 San Marcos 194 2017 Santa rosa 335 2017 Sololá 166 2017 Suchitepéquez 371 2017 Totonicapán 64 2017 Zacapa 264 2018 Alta Verapaz 355 2018 Baja Verapaz 190 2018 Chimaltenango 347 2018 Chiquimula 275 2018 El progreso 259 2018 Escuintla 629 2018 Guatemala 4054 2018 Huehuetenango 158 2018 Izabal 318 2018 Jalapa 120 2018 Jutiapa 241 2018 Peten 282 2018 Quetzaltenango 238 2018 Quiche 203 2018 Retalhuleu 245 2018 Sacatepéquez 297 2018 San Marcos 226 2018 Santa rosa 294 2018 Sololá 138 2018 Suchitepéquez 306 2018 Totonicapán 63 2018 Zacapa 276 2019 Alta Verapaz 453 2019 Baja Verapaz 196 2019 Chimaltenango 407 2019 Chiquimula 222 2019 El progreso 232 2019 Escuintla 699 2019 Guatemala 4699 2019 Huehuetenango 142 2019 Izabal 374 2019 Jalapa 141 2019 Jutiapa 336 2019 Peten 335 2019 Quetzaltenango 291 2019 Quiche 174 2019 Retalhuleu 329 2019 Sacatepéquez 363 2019 San Marcos 235 2019 Santa rosa 339 2019 Sololá 197 2019 Suchitepéquez 337 2019 Totonicapán 87 2019 Zacapa 239 2020 Alta Verapaz 470 2020 Baja Verapaz 153 2020 Chimaltenango 338 2020 Chiquimula 204 2020 El progreso 198 2020 Escuintla 718 2020 Guatemala 4307 2020 Huehuetenango 165 2020 Izabal 344 2020 Jalapa 186 2020 Jutiapa 258 2020 Peten 312 2020 Quetzaltenango 323 2020 Quiche 160 2020 Retalhuleu 340 2020 Sacatepéquez 323 2020 San Marcos 240 2020 Santa rosa 314 2020 Sololá 116 2020 Suchitepéquez 324 2020 Totonicapán 76 2020 Zacapa 234 2021 Alta Verapaz 580 2021 Baja Verapaz 212 2021 Chimaltenango 452 2021 Chiquimula 253 2021 El progreso 271 2021 Escuintla 986 2021 Guatemala 5352 2021 Huehuetenango 236 2021 Izabal 379 2021 Jalapa 230 2021 Jutiapa 369 2021 Peten 396 2021 Quetzaltenango 407 2021 Quiche 290 2021 Retalhuleu 361 2021 Sacatepéquez 438 2021 San Marcos 302 2021 Santa rosa 370 2021 Sololá 161 2021 Suchitepéquez 375 2021 Totonicapán 115 2021 Zacapa 261 2022 Alta Verapaz 528 2022 Baja Verapaz 201 2022 Chimaltenango 444 2022 Chiquimula 205 2022 El 

Databricks visualization. Run in Databricks to view.

Analizando las gráficas es evidente que el lugar con mayor cantidad de accentes con diferencia es la capital de Guatemala. Esto tiene sentido considerando que es aquí el lugar donde hay una mayor concentración de presonas y carros.

# Distribución de accidentes por día en 2023

In [0]:
from pyspark.sql import functions as F
from pyspark.sql.functions import col

# Verificar los valores reales en la columna
print("Valores únicos en dia_sem_ocu:")
hechos_transito.select("dia_sem_ocu").distinct().orderBy("dia_sem_ocu").show()

# Verificar cuántos registros tienen valores nulos en dia_sem_ocu para 2023
nulos_2023 = hechos_transito.filter(F.col("año") == 2023).filter(F.col("dia_sem_ocu").isNull()).count()
print(f"Registros con dia_sem_ocu nulo en 2023: {nulos_2023}")

# Corregir la transformación - comparar con valores double
hechos_transito = hechos_transito.withColumn(
    "dia_semana_desc",
    F.when(F.col("dia_sem_ocu") == 1.0, "Domingo")
    .when(F.col("dia_sem_ocu") == 2.0, "Lunes")
    .when(F.col("dia_sem_ocu") == 3.0, "Martes")
    .when(F.col("dia_sem_ocu") == 4.0, "Miércoles")
    .when(F.col("dia_sem_ocu") == 5.0, "Jueves")
    .when(F.col("dia_sem_ocu") == 6.0, "Viernes")
    .when(F.col("dia_sem_ocu") == 7.0, "Sábado")
    .otherwise("Desconocido")
)

# Contar accidentes por día de semana en 2023
accidentes_2023_por_dia = hechos_transito.filter(F.col("año") == 2020) \
    .groupBy("dia_semana_desc") \
    .count() \
    .orderBy("count", ascending=False)

print("Accidentes por día de la semana en 2023:")
accidentes_2023_por_dia.show()

print("GRÁFICA DE COLUMNAS - Accidentes por día de la semana en 2023")
display(accidentes_2023_por_dia)


Valores únicos en dia_sem_ocu:
+-----------+
|dia_sem_ocu|
+-----------+
|        1.0|
|        2.0|
|        3.0|
|        4.0|
|        5.0|
|        6.0|
|        7.0|
+-----------+

Registros con dia_sem_ocu nulo en 2023: 0
Muestra de datos transformados:
+-----------+---------------+
|dia_sem_ocu|dia_semana_desc|
+-----------+---------------+
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
|        7.0|         Sábado|
+-----------+---------------+
only showing top 20 

dia_semana_desc count Viernes 1180 Sábado 1055 Domingo 880 Jueves 873 Lunes 796 Martes 783 Miércoles 783

Databricks visualization. Run in Databricks to view.

En general se puede ver que los días viernes son los más peligrosos con respecto a la cantidad accidentes en el año 2023.

# Distribución de accidentes por hora en el día, en el municipio Guatemala

In [0]:
from pyspark.sql import functions as F

# solo guate
accidentes_guatemala = hechos_transito.filter(F.col("mupio_ocu") == 101.0)

print("Número de accidentes en municipio de Guatemala:")
accidentes_guatemala.count()

# Crear la columna de hora
distribucion_horas = accidentes_guatemala.withColumn(
    "hora", F.floor(F.col("hora_ocu"))
).filter(
    F.col("hora").between(0, 23)  # Filtrar horas válidas
).groupBy(
    "hora"
).agg(
    F.count("*").alias("cantidad_accidentes")
).orderBy(
    "hora"
)

# Mostrar los resultados
print("Distribución de accidentes por hora en municipio de Guatemala:")

display(distribucion_horas)

Número de accidentes en municipio de Guatemala:
Distribución de accidentes por hora en municipio de Guatemala:


hora cantidad_accidentes 0 737 1 734 2 575 3 472 4 370 5 417 6 452 7 559 8 512 9 587 10 659 11 571 12 602 13 529 14 614 15 718 16 626 17 820 18 853 19 967 20 989 21 1054 22 979 23 996

Databricks visualization. Run in Databricks to view.

Como se puede ver en la gráfica, es más peligroso en la noche. Parece que ha medida que avanza el día, aumenta la cantidad de accidentes.